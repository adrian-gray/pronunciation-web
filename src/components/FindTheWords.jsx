import React from 'react'

import {
  Paper,
  Typography,
  withStyles
} from '@material-ui/core'

const styles = (theme) => ({
  headspace: theme.headspace,
  largeText: theme.largeText,
  left: {
    ...theme.subCanvas,
    float: 'left',
    margin: '2rem',
    padding: '1rem'
  },
  clearFloat: {
    clear: 'both'
  }
})

const FindTheWords = (props) => {
  const { classes, tag, words } = props

  const wordList = words.map((word, index) => {
    return (
      <div key={index}>
        <Paper className={classes.left}>
          <Typography className={classes.largeText} gutterBottom>
            {word}
          </Typography>
        </Paper>
      </div>
    )
  })

  return (
    <div className={classes.headspace}>
      <Typography variant='title' gutterBottom>
        {`Find the words with a ${tag}`}
      </Typography>
      {wordList}
      <div className={classes.clearFloat} />
    </div>
  )
}

export default withStyles(styles)(FindTheWords)
