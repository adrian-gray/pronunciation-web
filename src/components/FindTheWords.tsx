import React, { useState } from "react";
import { IFindTheWordsProps } from "./../../@types/PronounceWeb";

import Container from "react-bootstrap/Container";

import MemberGate from "./MemberGate";
import SplitHilite from "./SplitHilite";
import Tile from "./Tile";

export default (props: IFindTheWordsProps) => {
  const [isCorrect, setIsCorrect] = useState(
    Array(props.words.length).fill(undefined)
  );
  const [allCorrect, setAllCorrect] = useState(false);

  function checkAllCorrect(correct: boolean[]) {
    let numRemaining = props.correct.length;

    for (let i = 0; i < props.words.length; i++) {
      if (correct[i] === true) {
        numRemaining--;
      } else if (correct[i] === false) {
        return false;
      }
    }
    return numRemaining === 0;
  }

  const handleClick = (index: number) => {
    const isWordCorrect = props.correct.includes(props.words[index]);
    const newStateIsCorrect = isCorrect.slice();

    newStateIsCorrect[index] = isWordCorrect;
    setAllCorrect(checkAllCorrect(newStateIsCorrect));
    setIsCorrect(newStateIsCorrect);
  };

  const { ipa, isHearTheDifference, tag, words } = props;

  let description = "";
  let result;
  let resultBG = false;
  let title;

  if (allCorrect) {
    result = <h3 className="text-center">Yes âœ“</h3>;
    resultBG = true;
  }

  const wordTiles = words.map((word: string, index: number) => (
    <Tile
      word={word}
      key={index}
      index={index}
      isCorrect={isCorrect[index]}
      handleClick={handleClick}
    />
  ));

  if (isHearTheDifference) {
    title = (
      <h3>
        {`Odd One Out - select words that DON'T have an `}
        <SplitHilite str={ipa} />
        {`sound`}
      </h3>
    );
    description = `Find the Odd One Out. Say the words and select those without the ${tag} vowel sound. Once you get a word, it is replaced with a more challenging word. Can you find all the odd words before your time runs out? You can click on words to hear the pronunciation. Odd One Out helps you recognise and pronounce the vowel sounds of common English words.`;
  } else {
    title = (
      <h3>
        {`Select the words with an `}
        <SplitHilite str={ipa} />
        {`sound`}
      </h3>
    );
    description = `Find the correct sound. Say the words and select those with the ${tag} vowel sound. Once you get a word, it is replaced with a more challenging word. Can you find all the words before your time runs out? You can click on words to hear the pronunciation. Find the words helps you recognise and pronounce the vowel sounds of common English words.`;
  }

  const display = (
    <Container className={resultBG ? "correct" : "incorrect"}>
      {wordTiles}
      <div className="clear-float" />
      {result}
    </Container>
  );

  return (
    <Container>
      {title}
      <p>{description}</p>
      <MemberGate content={display} isUserAuth={props.isUserAuth} />
    </Container>
  );
};
