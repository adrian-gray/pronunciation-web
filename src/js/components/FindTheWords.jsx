import React from 'react'

export default function FindTheWords (props) {
  const words = props.words.map((word, index) => {
    return <div key={index} className='col'>{word}</div>
  })

  return (
    <div className='container'>
      <div className='row'>
        {words}
      </div>
    </div>
  )
}
