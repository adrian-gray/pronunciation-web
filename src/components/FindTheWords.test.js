import React from "react";
import { render, cleanup } from "@testing-library/react";

import FindTheWords from "./FindTheWords";

afterEach(cleanup);

test("FindTheWords", async () => {
  const props = {
    ipa: "a:",
    words: ["foo", "bar", "baz", "quox", "apple", "banana", "chad", "elle"],
    correct: ["foo", "bar", "baz"],
    userAuth: 1
  };

  const { container } = render(<FindTheWords {...props} />);

  expect(container).toMatchSnapshot();
});
