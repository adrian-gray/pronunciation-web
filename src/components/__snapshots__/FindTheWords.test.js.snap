// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FindTheWords 1`] = `
<div>
  <div
    class="headspace"
  >
    <h3>
      Select the words with an 
      <span>
        a:
      </span>
       
      sound
    </h3>
    <p>
      Find the correct sound. Say the words and select those with the ~a vowel sound. Once you get a word, it is replaced with a more challenging word. Can you find all the words before your time runs out? You can click on words to hear the pronunciation. Find the words helps you recognise and pronounce the vowel sounds of common English words.
    </p>
    <div
      class="incorrect"
    >
      <div
        class="cursor-pointer"
      >
        <div
          class="tile"
        >
          <p
            class="large-text"
          >
            <span>
              foo
            </span>
          </p>
        </div>
      </div>
      <div
        class="cursor-pointer"
      >
        <div
          class="tile"
        >
          <p
            class="large-text"
          >
            <span>
              bar
            </span>
          </p>
        </div>
      </div>
      <div
        class="cursor-pointer"
      >
        <div
          class="tile"
        >
          <p
            class="large-text"
          >
            <span>
              baz
            </span>
          </p>
        </div>
      </div>
      <div
        class="cursor-pointer"
      >
        <div
          class="tile"
        >
          <p
            class="large-text"
          >
            <span>
              quox
            </span>
          </p>
        </div>
      </div>
      <div
        class="cursor-pointer"
      >
        <div
          class="tile"
        >
          <p
            class="large-text"
          >
            <span>
              apple
            </span>
          </p>
        </div>
      </div>
      <div
        class="cursor-pointer"
      >
        <div
          class="tile"
        >
          <p
            class="large-text"
          >
            <span>
              banana
            </span>
          </p>
        </div>
      </div>
      <div
        class="cursor-pointer"
      >
        <div
          class="tile"
        >
          <p
            class="large-text"
          >
            <span>
              chad
            </span>
          </p>
        </div>
      </div>
      <div
        class="cursor-pointer"
      >
        <div
          class="tile"
        >
          <p
            class="large-text"
          >
            <span>
              elle
            </span>
          </p>
        </div>
      </div>
      <div
        class="clear-float"
      />
    </div>
  </div>
</div>
`;
